### PR TITLE
Add Arrays library with unit tests (#1209)

### PR DESCRIPTION
**Pull request by jbogacz**
* prepared due to snapshot token requirements
* add library with method to find upper bound
* add unit test for basic and edge cases

Blocks PR #1331

# 🚀 Description

Arrays library which is the requirement for ERC20 token with snapshots. Blocks PR #1331. 

- [+] 📘 I've reviewed the [OpenZeppelin Contributor Guidelines](../blob/master/CONTRIBUTING.md)
- [+] ✅ I've added tests where applicable to test my new functionality.
- [+] 📖 I've made sure that my contracts are well-documented.
- [+] 🎨 I've run the JS/Solidity linters and fixed any issues (`npm run lint:fix`).

### CODE REVIEW

**Comment by jbogacz** [comment]
@frangio @nventuro hey guys! I'm not sure if does it stick to general concept but I tried. Let me know if you have any comments are ideas for missing test cases.

File: contracts/utils/Arrays.sol [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+
+/**
+ * @title Arrays
+ * @dev Utility library of inline array functions
+ */
+library Arrays {
+
+  /**
+   * @dev Find upper bound for searching element. Utilize binary search function with O(log(n)) cost.
```
Comment (nventuro) [inline contracts/utils/Arrays.sol]: We should improve this documentation: in particular, it is critical to document that `element` is the value we're searching for, that _array_ is sorted in ascending order, and the return value is an _index_.

We should also better explain what the function does: given said sorted array, it returns the smallest array index such that the value at that index is larger or equal to the received element, or `array.length` if no such index exists (i.e. all values inside the array are smaller that the target). It also returns 0 for an empty array, though we may want to revert in that scenario.

File: test/utils/Arrays.test.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      this.arrays = await ArraysImpl.new(WITH_GAP_ARRAY);
+    });
+
+    it('should return `last index + 1` of first range for the element from the gap', async function () {
+      (await this.arrays.findUpperBound(100)).should.be.bignumber.equal(65);
+    });
+  });
+
+  function range (start, end) {
+    return Array(end - start + 1).fill().map((_, idx) => start + idx);
```
Comment (nventuro) [inline test/utils/Arrays.test.js]: This implementation might not match people's expectations, since [`lodash`'s `range` doesn't include `end`](https://lodash.com/docs/4.17.10#range). Could we also add a short comment describing what it does?

File: test/utils/Arrays.test.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    beforeEach(async function () {
+      this.arrays = await ArraysImpl.new(ODD_ELEMENTS_ARRAY);
+    });
+
+    it('should return correct index for the basic case', async function () {
+      (await this.arrays.findUpperBound(139)).should.be.bignumber.equal(11);
+    });
+
+    it('should return `index 0` for the first element', async function () {
```
Comment (nventuro) [inline test/utils/Arrays.test.js]: I'd change these to just 'should return 0 for the first element'

File: test/utils/Arrays.test.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      (await this.arrays.findUpperBound(128)).should.be.bignumber.equal(0);
+    });
+
+    it('should return `last index` for the last element', async function () {
+      (await this.arrays.findUpperBound(256)).should.be.bignumber.equal(128);
+    });
+
+    it('should return `last index + 1` for the element over the upper boundary', async function () {
+      (await this.arrays.findUpperBound(512)).should.be.bignumber.equal(129);
```
Comment (nventuro) [inline test/utils/Arrays.test.js]: The fact that the array is 129 elements long, and the first element is 128, may be a bit confusing. Consider using e.g. a shorter array.

File: test/utils/Arrays.test.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  });
+
+  context('Array with gap', function () {
+    const WITH_GAP_ARRAY = [...range(0, 64), ...range(128, 256)];
+
+    beforeEach(async function () {
+      this.arrays = await ArraysImpl.new(WITH_GAP_ARRAY);
+    });
+
+    it('should return `last index + 1` of first range for the element from the gap', async function () {
```
Comment (nventuro) [inline test/utils/Arrays.test.js]: I _really_ like what you're testing for here, but fear it may be hard to follow, due to the arrays being generated by `range`, and there being so many magic numbers. What if we simply made a short test array? e.g. `const gapArray = [0, 1, 2, 3,4, 8, 9, 10, 11]`, and then test for `7`.

File: test/utils/Arrays.test.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
+const ArraysImpl = artifacts.require('ArraysImpl');
+
+const BigNumber = web3.BigNumber;
+
+require('chai')
+  .use(require('chai-bignumber')(BigNumber))
+  .should();
+
+contract('Arrays', function () {
+  context('Odd number of elements', function () {
```
Comment (nventuro) [inline test/utils/Arrays.test.js]: Love the testing for odd and even elements! We should also add a test for an empty array (and decide what to do in that scenario).

File: test/utils/Arrays.test.js (hunk 5 above)
Comment (nventuro) [reply test/utils/Arrays.test.js]: We may also want to use a shorter range in all tests, actually, to make it easier for the reader to understand what's going on.

File: contracts/utils/Arrays.sol [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  ) 
+    internal
+    view
+    returns (uint256)
+  {
+    uint256 low = 0;
+    uint256 high = array.length;
+
+    while (low < high) {
+      uint256 mid = Math.average(low, high);
```
[the remark this answers is not in the document]
Comment (jbogacz) [reply contracts/utils/Arrays.sol]: @nventuro I'm sorry but I really don't see the point here :( Could you provide more context to where it's possible to access array as `array[array.length]' ?

[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/utils/Arrays.sol]: It isn't, because of how `Math.average` works, that's my point :grin:

[the remark this answers is not in the document]
Comment (jbogacz) [reply contracts/utils/Arrays.sol]: Hah.. ok you got me :laughing: So I will try to put some comment here too

[the remark this answers is not in the document]
Comment (jbogacz) [reply contracts/utils/Arrays.sol]: Or you can propose one.. and make my life better :smiley:

[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/utils/Arrays.sol]: Haha sorry! Something like this should work:

```
Note that mid will always be strictly less than high (i.e. it will be a valid array index) because Math.average rounds down (it does integer division with truncation).
```

[the remark this answers is not in the document]
Comment (jbogacz) [reply contracts/utils/Arrays.sol]: Good thanks! I could never do it better! 😄 I see it's last missing thing for now so I will push it shortly. The last time - @nventuro Thanks a lot for that!

**Comment by jbogacz** [comment]
Nice, just want to thanks again for your help here 😄 I will go for #1331 now.
